File: 41-first-missing-positive/first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        int i = 0;
        while (i < n) {
            if (nums[i] <= 0 || nums[i] > n) {
                i++;
            } else {
                int correctindex = nums[i] - 1;
                if (nums[i] > 0 && nums[i] <= n &&
                    nums[i] != nums[correctindex]) {
                        swap(nums[i], nums[correctindex]);
                    // int temp = nums[i];
                    // nums[i] = nums[correctindex];
                    // nums[correctindex] = temp;
                } else {
                    i++;
                }
            }
        }

        for (int j = 0; j < n; j++) {
            if (nums[j] != j + 1) {
                return j + 1;
            }
        }
        return n + 1;
    }
};

File: 41-first-missing-positive/first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        i = 0

        while i < n:
            correct_index = nums[i] - 1

            if 1 <= nums[i] <= n and nums[i] != nums[correct_index]:
                nums[i], nums[correct_index] = nums[correct_index], nums[i]
            else:
                i += 1

        for j in range(n):
            if nums[j] != j + 1:
                return j + 1

        return n + 1

